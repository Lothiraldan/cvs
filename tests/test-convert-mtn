#!/bin/sh

"$TESTDIR/hghave" mtn || exit 80

# Monotone directory is called .monotone on *nix and monotone
# on Windows. Having a variable here ease test patching.
mtndir=.monotone
echo "[extensions]" >> $HGRCPATH
echo "convert=" >> $HGRCPATH
echo 'hgext.graphlog =' >> $HGRCPATH

HOME=`pwd`/do_not_use_HOME_mtn; export HOME
# Windows version of monotone home
APPDATA=$HOME; export APPDATA

echo % tedious monotone keys configuration
# The /dev/null redirection is necessary under Windows, or
# it complains about home directory permissions
mtn --quiet genkey test@selenic.com 1>/dev/null 2>&1 <<EOF
passphrase
passphrase
EOF
cat >> $HOME/$mtndir/monotonerc <<EOF
function get_passphrase(keypair_id)
    return "passphrase"
end
EOF

echo % create monotone repository
mtn db init --db=repo.mtn
mtn --db=repo.mtn --branch=com.selenic.test setup workingdir
cd workingdir
echo a > a
mkdir dir
echo b > dir/b
echo d > dir/d
python -c 'file("bin", "wb").write("a\\x00b")'
echo c > c
mtn add a dir/b dir/d c bin
mtn ci -m initialize
echo % update monotone working directory
mtn mv a dir/a
echo a >> dir/a
echo b >> dir/b
mtn drop c
python -c 'file("bin", "wb").write("b\\x00c")'
mtn ci -m update1
cd ..

echo % convert once
hg convert -s mtn repo.mtn

cd workingdir
echo e > e
mtn add e
mtn drop dir/b
mtn mv bin bin2
mtn ci -m 'update2 "with" quotes'
echo '% test directory move'
mkdir -p dir1/subdir1
mkdir -p dir1/subdir2_other
echo file1 > dir1/subdir1/file1
echo file2 > dir1/subdir2_other/file1
mtn add dir1/subdir1/file1 dir1/subdir2_other/file1
mtn ci -m createdir1
mtn rename dir1/subdir1 dir1/subdir2
mtn ci -m movedir1
echo '% test subdirectory move'
mtn mv dir dir2
echo newfile > dir2/newfile
mtn drop dir2/d
mtn add dir2/newfile
mtn ci -m movedir
# Test directory removal with empty directory
mkdir dir2/dir
mkdir dir2/dir/subdir
echo f > dir2/dir/subdir/f
mkdir dir2/dir/emptydir
mtn add --quiet -R dir2/dir
mtn ci -m emptydir
mtn drop -R dir2/dir
mtn ci -m dropdirectory
echo '% test directory and file move'
mkdir -p dir3/d1
echo a > dir3/a
mtn add dir3/a dir3/d1
mtn ci -m dirfilemove
mtn mv dir3/a dir3/d1/a
mtn mv dir3/d1 dir3/d2
mtn ci -m dirfilemove2
cd ..

echo % convert incrementally
hg convert -s mtn repo.mtn

glog()
{
    hg glog --template '#rev# "#desc|firstline#" files: #files#\n' "$@"
}

cd repo.mtn-hg
hg up -C
glog
echo % manifest
hg manifest
echo % contents
cat dir2/a
test -d dir2/dir && echo 'removed dir2/dir is still there!'

echo % file move
hg log -v -C -r 1 | grep copies
echo % check directory move
hg manifest -r 4
test -d dir1/subdir2 || echo 'new dir1/subdir2 does not exist!'
test -d dir1/subdir1 && echo 'renamed dir1/subdir1 is still there!'
hg log -v -C -r 4 | grep copies
echo % check file remove with directory move
hg manifest -r 5
echo % check file move with directory move
hg manifest -r 9
exit 0

