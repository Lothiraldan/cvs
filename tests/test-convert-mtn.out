% tedious monotone keys configuration
% create monotone repository
mtn: adding a to workspace manifest
mtn: adding bin to workspace manifest
mtn: adding c to workspace manifest
mtn: adding dir to workspace manifest
mtn: adding dir/b to workspace manifest
mtn: adding dir/d to workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision 0f6e5e4f2e7d2a8ef312408f57618abf026afd90
% update monotone working directory
mtn: skipping dir, already accounted for in workspace
mtn: renaming a to dir/a in workspace manifest
mtn: dropping c from workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision 51d0a982464573a2a2cf5ee2c9219c652aaebeff
% convert once
assuming destination repo.mtn-hg
initializing destination repo.mtn-hg repository
scanning source...
sorting...
converting...
1 initialize
0 update1
mtn: adding e to workspace manifest
mtn: dropping dir/b from workspace manifest
mtn: renaming bin to bin2 in workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision ebe58335d85d8cb176b6d0a12be04f5314b998da
% test directory move
mtn: adding dir1 to workspace manifest
mtn: adding dir1/subdir1 to workspace manifest
mtn: adding dir1/subdir1/file1 to workspace manifest
mtn: adding dir1/subdir2_other to workspace manifest
mtn: adding dir1/subdir2_other/file1 to workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision a8d62bc04fee4d2936d28e98bbcc81686dd74306
mtn: skipping dir1, already accounted for in workspace
mtn: renaming dir1/subdir1 to dir1/subdir2 in workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision 2c3d241bbbfe538b1b51d910f5676407e3f4d3a6
% test subdirectory move
mtn: renaming dir to dir2 in workspace manifest
mtn: dropping dir2/d from workspace manifest
mtn: adding dir2/newfile to workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision fdb5a02dae8bfce3a79b3393680af471016e1b4c
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision 8bbf76d717001d24964e4604739fdcd0f539fc88
mtn: dropping dir2/dir/subdir/f from workspace manifest
mtn: dropping dir2/dir/subdir from workspace manifest
mtn: dropping dir2/dir/emptydir from workspace manifest
mtn: dropping dir2/dir from workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision 2323d4bc324e6c82628dc04d47a9fd32ad24e322
% test directory and file move
mtn: adding dir3 to workspace manifest
mtn: adding dir3/a to workspace manifest
mtn: adding dir3/d1 to workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision 47b192f720faa622f48c68d1eb075b26d405aa8b
mtn: skipping dir3/d1, already accounted for in workspace
mtn: renaming dir3/a to dir3/d1/a in workspace manifest
mtn: skipping dir3, already accounted for in workspace
mtn: renaming dir3/d1 to dir3/d2 in workspace manifest
mtn: beginning commit on branch 'com.selenic.test'
mtn: committed revision 8b543a400d3ee7f6d4bb1835b9b9e3747c8cb632
% convert incrementally
assuming destination repo.mtn-hg
scanning source...
sorting...
converting...
7 update2 "with" quotes
6 createdir1
5 movedir1
4 movedir
3 emptydir
2 dropdirectory
1 dirfilemove
0 dirfilemove2
7 files updated, 0 files merged, 0 files removed, 0 files unresolved
@  9 "dirfilemove2" files: dir3/a dir3/d2/a
|
o  8 "dirfilemove" files: dir3/a
|
o  7 "dropdirectory" files: dir2/dir/subdir/f
|
o  6 "emptydir" files: dir2/dir/subdir/f
|
o  5 "movedir" files: dir/a dir/d dir2/a dir2/newfile
|
o  4 "movedir1" files: dir1/subdir1/file1 dir1/subdir2/file1
|
o  3 "createdir1" files: dir1/subdir1/file1 dir1/subdir2_other/file1
|
o  2 "update2 "with" quotes" files: bin bin2 dir/b e
|
o  1 "update1" files: a bin c dir/a dir/b
|
o  0 "initialize" files: a bin c dir/b dir/d

% manifest
bin2
dir1/subdir2/file1
dir1/subdir2_other/file1
dir2/a
dir2/newfile
dir3/d2/a
e
% contents
a
a
% file move
copies:      dir/a (a)
% check directory move
bin2
dir/a
dir/d
dir1/subdir2/file1
dir1/subdir2_other/file1
e
copies:      dir1/subdir2/file1 (dir1/subdir1/file1)
% check file remove with directory move
bin2
dir1/subdir2/file1
dir1/subdir2_other/file1
dir2/a
dir2/newfile
e
% check file move with directory move
bin2
dir1/subdir2/file1
dir1/subdir2_other/file1
dir2/a
dir2/newfile
dir3/d2/a
e
